lib/valueflow.cpp

```diff
@@ -1658,8 +1658,79 @@ static bool evalAssignment(ValueFlow::Value &lhsValue, const std::string &assign
     return true;
 }
 
+template<class T>
+struct SingleRange {
+    T* x;
+    T* begin() const {
+        return x;
+    }
+    T* end() const {
+        return x+1;
+    }
+};
+
+template<class T>
+SingleRange<T> MakeSingleRange(T& x)
+{
+    return {&x};
+}
+
+class MultiRange {
+    using M = std::unordered_map<nonneg int, ValueFlow::Value>;
+
+    struct Iterator {
+        using value_type = const ValueFlow::Value;
+        using pointer = value_type *;
+        using reference = value_type &;
+
+        explicit Iterator(const M::const_iterator &it)
+            : mIt(it) {
+        }
+
+        reference operator*() const {
+            return mIt->second;
+        }
+
+        pointer operator->() {
+            return &mIt->second;
+        }
+
+        Iterator &operator++() {
+            ++mIt;
+            return *this;
+        }
+
+        friend bool operator==(const Iterator &a, const Iterator &b) {
+            return a.mIt == b.mIt;
+        }
+
+        friend bool operator!=(const Iterator &a, const Iterator &b) {
+            return a.mIt != b.mIt;
+        }
+
+    private:
+        M::const_iterator mIt;
+    };
+
+public:
+    explicit MultiRange(const M &m)
+        : mMap(m) {
+    }
+
+    Iterator begin() const {
+        return Iterator(mMap.begin());
+    }
+    Iterator end() const {
+        return Iterator(mMap.end());
+    }
+
+private:
+    const M &mMap;
```

Comment:
It should store it by pointer, so the class can be regular.

Reply:
done